### PR TITLE
Add timeout for cluster ready in osd_create.sh

### PR DESCRIPTION
**Pull request by asergienk**
Currently script waits forever for the cluster to be ready, adding a 90 minutes timeout will save us time in retry.

### CODE REVIEW

File: ci-operator/step-registry/osd/create/create/osd-create-create-commands.sh [hunk 1]
```diff
@@ -115,10 +116,15 @@ echo '{ "expiration_timestamp": "'"${INIT_EXPIRATION_DATE}"'" }' | ocm patch "/a
...
 echo "Waiting for cluster ready..."
+start_time=$(date +"%s")
 while true; do
   sleep 60
   CLUSTER_STATE=$(ocm get cluster "${CLUSTER_ID}" | jq -r '.status.state')
   echo "Cluster state: ${CLUSTER_STATE}"
+  if (( $(date +"%s") - $start_time >= $CLUSTER_TIMEOUT )); then
+    echo "error: Timed out while waiting for cluster to be ready"
+    exit 1
+  fi
```
Comment (lslebodn) [inline ci-operator/step-registry/osd/create/create/osd-create-create-commands.sh:127]: I would move this if block after the check `[[ "${CLUSTER_STATE}" == "ready" ]]`
Just in case we hit limit CLUSTER_TIMEOUT (1.5 hour ATM) but cluster is finally ready :-)

**Comment by jupierce** [comment]
Shouldn't this be done with step metadata: https://docs.ci.openshift.org/docs/architecture/timeouts/#step-registry-test-process-timeouts ?

File: ci-operator/step-registry/osd/create/create/osd-create-create-commands.sh (hunk 1 above)
Comment (asergienk) [reply ci-operator/step-registry/osd/create/create/osd-create-create-commands.sh:127]: Good point, thanks! I'll change it now

**Comment by samvarankashyap** [comment]
@jupierce Great suggestion ! 
According to your comment, 
We have added timeout at both step level and also as a parameter because we would like to give the customer the ability to override the timeout for osd clusters.
 Please check and let us know what you think. 
 Thanks.

File: ci-operator/step-registry/osd/create/create/osd-create-create-ref.yaml [hunk 2]
```diff
@@ -7,6 +7,7 @@ ref:
     requests:
       cpu: 100m
       memory: 300Mi
+  timeout: 3h0m0s
```
Comment (nmars) [inline ci-operator/step-registry/osd/create/create/osd-create-create-ref.yaml:10]: Why 3 hours here? Could we make this 1h?

Comment (samvarankashyap) [reply ci-operator/step-registry/osd/create/create/osd-create-create-ref.yaml:10]: @asergienk could you make it 2hrs.

Comment (asergienk) [reply ci-operator/step-registry/osd/create/create/osd-create-create-ref.yaml]: done, it's 2hrs now

File: ci-operator/step-registry/osd/create/create/osd-create-create-commands.sh [hunk 3]
```diff
@@ -19,6 +19,7 @@ CLUSTER_MULTI_AZ=${CLUSTER_MULTI_AZ:-false}
 SSO_CLIENT_ID=$(read_profile_file "sso-client-id")
 SSO_CLIENT_SECRET=$(read_profile_file "sso-client-secret")
 OCM_TOKEN=$(read_profile_file "ocm-token")
+CLUSTER_TIMEOUT=${CLUSTER_TIMEOUT:-}
```
Comment (lslebodn) [inline ci-operator/step-registry/osd/create/create/osd-create-create-commands.sh:22]: `:-}` means setting default value for empty string.
But empty string will cause syntax failure in the `if`

```
bash-5.1 ~$ unset CLUSTER_TIMEOUT
bash-5.1 ~$ CLUSTER_TIMEOUT=${CLUSTER_TIMEOUT:-}
bash-5.1 ~$ start_time=$(date +"%s")
bash-5.1 ~$ (( $(date +"%s") - $start_time >= $CLUSTER_TIMEOUT ))
bash: ((: 1651045178 - 1651045174 >=  : syntax error: operand expected (error token is ">=  ")
```

Value should be always passed to the script  based on `osd-create-create-ref.yaml`
IMHO setting default value to empty string is not very useful we should rather rely on strict bash options `set -o nounset` to report unset variable

```
bash-5.1 ~$ cat test.sh 

#!/bin/bash
set -o nounset
set -o errexit
set -o pipefail

CLUSTER_TIMEOUT=${CLUSTER_TIMEOUT}

echo "Waiting for cluster ready..."
start_time=$(date +"%s")

if (( $(date +"%s") - $start_time >= $CLUSTER_TIMEOUT )); then
     echo "error: Timed out while waiting for cluster to be ready"
     exit 1
fi
bash-5.1 ~$ bash test.sh
test.sh: line 7: CLUSTER_TIMEOUT: unbound variable

bash-5.1 ~$ CLUSTER_TIMEOUT=10 bash test.sh
Waiting for cluster ready...
bash-5.1 ~$ echo $?
0
```

Comment (asergienk) [reply ci-operator/step-registry/osd/create/create/osd-create-create-commands.sh]: thanks for such detailed explanation! It totally make sense, I fixed it based on your comment.

**Comment by bradmwilliams** [comment]
> @bradmwilliams @jupierce could you ptal? Thanks!

While I think these changes look OK, I'm not quite sure I quite understand the need for it.  In my experience, the success of getting a cluster up is more important that shaving a few minutes out of the process. 

If OSD is anything like the other cloud platforms, where the installation time is quite variable, time boxing it seems kind of risky.

File: ci-operator/step-registry/osd/create/create/osd-create-create-ref.yaml [hunk 4]
```diff
@@ -31,6 +32,9 @@ ref:
   - name: CLUSTER_MULTI_AZ
     default: "false"
     documentation: Set to 'true' if you want to deploy a cluster across multiple availability zones.
+  - name: CLUSTER_TIMEOUT
+    default: "5400"
```
Comment (nmars) [inline ci-operator/step-registry/osd/create/create/osd-create-create-ref.yaml:36]: Just a question/observation. The `timeout` in line 10 is set to `2h0m0s`. The `CLUSTER_TIMEOUT` environment variable's `default` is set to `5400`. I assume that's seconds. Which is 1.5 hours.

How are these properties related?

Comment (asergienk) [reply ci-operator/step-registry/osd/create/create/osd-create-create-ref.yaml:36]: They are not related. The ```timeout``` in line 10 is a timeout for the osd cluster create step.
The ```CLUSTER_TIMEOUT``` variable is a timeout for the cluster to obtain the status ```ready```. If the cluster is not ```ready``` in 5400 seconds (1.5 hours) the execution stops.

Comment (samvarankashyap) [reply ci-operator/step-registry/osd/create/create/osd-create-create-ref.yaml:36]: > > @bradmwilliams @jupierce could you ptal? Thanks!
> 
> While I think these changes look OK, I'm not quite sure I quite understand the need for it. In my experience, the success of getting a cluster up is more important that shaving a few minutes out of the process.
> 
> If OSD is anything like the other cloud platforms, where the installation time is quite variable, time boxing it seems kind of risky.

@bradmwilliams 
I totally agree with your statement of timeboxing being risky approach. However, Within CVP MSP effort we would not want to wait for more cluster creation for beyond certain time. We would rather create a new cluster. 
If adding default values is a concern, we can always set to null value and based on the parameter passed to the step we can customize script to use the CLUSTER_TIMEOUT or not. 
If CLUSTER_TIMEOUT is null we would not timeout at all. 
Would that work ?

Comment (bradmwilliams) [reply ci-operator/step-registry/osd/create/create/osd-create-create-ref.yaml:36]: Right now, the step timeout is set to: `timeout: 2h0m0s` (which is the default value for steps anyway).  I'd recommend making the default `CLUSTER_TIMEOUT` value `7200`.  This way you are not changing any existing behavior.  Then, in your specific jobs, you can tweak the `CLUSTER_TIMEOUT` values to your liking.

Comment (samvarankashyap) [reply ci-operator/step-registry/osd/create/create/osd-create-create-ref.yaml:36]: @bradmwilliams we can set it to 7200

Comment (asergienk) [reply ci-operator/step-registry/osd/create/create/osd-create-create-ref.yaml]: it is set to 7200, @bradmwilliams could you ptal?

**Review by bradmwilliams — APPROVED** [review]
